hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSource.java

```diff
@@ -792,9 +795,13 @@ protected void shipEdits(WALEntryBatch entryBatch) {
     }
 
     private void updateLogPosition(long lastReadPosition) {
-      manager.logPositionAndCleanOldLogs(lastLoggedPath, peerClusterZnode, lastReadPosition,
-        this.replicationQueueInfo.isQueueRecovered(), false);
-      lastLoggedPosition = lastReadPosition;
+      try {
+        manager.logPositionAndCleanOldLogs(lastLoggedPath, peerClusterZnode, lastReadPosition,
+          this.replicationQueueInfo.isQueueRecovered(), false);
+        lastLoggedPosition = lastReadPosition;
+      } catch (ReplicationSourceWithoutPeerException re) {
+        source.terminate("Replication peer is removed and source should terminate", re);
```

Comment:
I am not an expert in this so please correct me if I missed anything.

Whenever zk listener is invoked for peersRemoved event, we do the following things.
https://github.com/apache/hbase/blob/branch-1/hbase-client/src/main/java/org/apache/hadoop/hbase/replication/ReplicationTrackerZKImpl.java#L175
https://github.com/apache/hbase/blob/branch-1/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceManager.java#L637-L640

One of the steps in ReplicationSourceManager#peerRemoved code path is to terminate the source but in addition to that it does many things like removingPeersFromHFileRefs and removing that source from ReplicationSourceManager#sources and ReplicationSourceManager#oldSources data structure. If we invoke the source#terminate method directly we will miss all these cleanup steps.

@apurtell  While typing this comment it looks like this patch got merged but I think we need more changes on this patch. 
Cc @sandeepvinayak  @wchevreuil

Reply:
thanks @shahrs87 , I am taking a look again if this can be improved.

Reply:
@shahrs87 To answer your concern, ref files get cleaned as part of replication zk node cleaner https://github.com/apache/hbase/blob/branch-1/hbase-server/src/main/java/org/apache/hadoop/hbase/master/cleaner/ReplicationZKNodeCleaner.java#L117

However, the source manager will have a reference to the terminated source which seems harmless but still should be removed as it will be the right thing to do, I can create the PR for this change.